in_toto/verifylib.py

```diff
@@ -635,6 +637,113 @@ def verify_all_item_rules(items, links, target_links=None):
         link.products, target_links)
 
 
+def verify_threshold_equality(layout, chain_link_dict):
+  """
+  <Purpose>
+    Iteratively verifies threshold conditions (if a step has been
+    performed and signed by enough number of functionaries) and link
+    artifacts (materials and products of a step is same for different
+    fucntionaries) of passed dictionary (chain_link_dict).
+
+  <Arguments>
+    layout:
+            The layout specified by the project owner against which the
+            threshold will be verified.
+
+    chain_link_dict:
+            A dictionary of key-link pair with step names as keys. For each
+            step name, there are one or more keyids and corresponding
+            link objects.
+
+  <Exceptions>
+    raises an Exception if threshold is not verified.
+    ThresholdVerificationError if the step is not performed by enough
+    functionaries or if the materials and products for a step are not same
+    for all functionaries.
+
+  <Side Effects>
+    None.
+
+  """
+
+  # We are only interested in links that are related to steps defined in the
+  # Layout, so iterate over layout.steps
+  for step in layout.steps:
+    # Skip steps that don't require multiple functionaries
+    if step.threshold <= 1:
+      continue
+
+    # Extract the key_link_dict for this step from the passed chain_link_dict
+    key_link_dict = chain_link_dict[step.name]
+
+    # take one exemplary link (e.g. the first in the step_link_dict)
+    compare_link = key_link_dict.values()[0]
+
+    # Check if we have at least <threshold> links for this step
+    if len(key_link_dict) < step.threshold:
+      raise ThresholdVerificationError("Step not performed by enough functionaries!".format())
+
+    # iterate over the remaining links to compare their properties
+    for keyid, link in six.iteritems(key_link_dict):
+      keys_dict = {}
+
+      # check if the authorized functionary has signed the link
+      for key, key_link in six.iteritems(layout.keys):
+        if key == keyid:
+          keys_dict[key] = key_link
+          verify_link_signatures(link, keys_dict)
+          break
```

Comment:
So what I do (or was trying to do) is, I extract the link object from the chain_link_dict.
Then, I extract the entire key object corresponding to the key in the filename.
Then I pass them to verify_link_signatures(link, keys_dict). This ensures that the link object has been indeed signed by the same key as the filename.

For eg. if we have foo.[Bob's keyid].link and foo.[Carl's keyid].link, and someone duplicates legitimate foo.[Bob's keyid].link by renaming it to foo.[Carl's keyid].link. We basically check if foo.[Carl's keyid].link was indeed signed by Carl's key.
Note that I am passing only a single key object as keys_dict in verify_link_signatures(link, keys_dict) instead of all the keys. This restricts the possibility of the link object being signed by any other key.